apps/emqx_durable_storage/test/emqx_ds_new_streams_SUITE.erl

```diff
@@ -0,0 +1,185 @@
+
+t_sub_unsub(_) ->
+    ?check_trace(
+        #{timetrap => 30_000},
+        begin
+            DB = ?FUNCTION_NAME,
+            {ok, Pid} = emqx_ds_new_streams:start_link(DB),
+            ?assertEqual(Pid, emqx_ds_new_streams:where(DB)),
+            %% Try to create subscriptions with invalid topic filter,
+            %% make sure the server handles that gracefully:
+            {error, badarg} = emqx_ds_new_streams:watch(DB, garbage),
+            %% Create a watch and make sure it is functional:
+            {ok, Ref} = emqx_ds_new_streams:watch(DB, [<<"foo">>, <<"1">>]),
+            ok = emqx_ds_new_streams:notify_new_stream(DB, [<<"foo">>, <<"1">>]),
+            assertEvents([Ref]),
+            %% Try to unsubscribe with wrong subid:
+            false = emqx_ds_new_streams:unwatch(DB, wrong),
+            %% Unsubscribe from the events and verify that notifications are
+            %% no longer received:
+            true = emqx_ds_new_streams:unwatch(DB, Ref),
+            ok = emqx_ds_new_streams:notify_new_stream(DB, [<<"foo">>, <<"1">>]),
+            assertEvents([])
+        end,
+        fun no_unexpected_events/1
+    ).
+
+%% Verify that subscriptions are automatically cleaned up when the
+%% subscribing process dies:
+t_clean_on_down(_) ->
+    ?check_trace(
+        #{timetrap => 10_000},
+        begin
+            DB = ?FUNCTION_NAME,
+            {ok, _} = emqx_ds_new_streams:start_link(DB),
+            %% Subscribe to topic updates from a temporary process:
+            Pid = spawn_link(
+                fun() ->
+                    {ok, _} = emqx_ds_new_streams:watch(DB, [<<"1">>]),
+                    {ok, _} = emqx_ds_new_streams:watch(DB, [<<"2">>]),
+                    receive
+                        done -> ok
+                    end
+                end
+            ),
+            %% Check if the subscriptions are present:
+            timer:sleep(100),
+            ?assertMatch([_, _], emqx_ds_new_streams:list_subscriptions(DB)),
+            %% Stop the process and verify that subscriptions were
+            %% automatically deleted:
+            Pid ! done,
+            timer:sleep(100),
+            ?assertMatch([], emqx_ds_new_streams:list_subscriptions(DB))
+        end,
+        fun no_unexpected_events/1
+    ).
+
+%% Verify that SUT is capable of forwarding notifications about
+%% changes in a group of topics (denoted by a topic filter) to a set
+%% of subscribers that also use topic filter:
+t_matching(_) ->
+    ?check_trace(
+        #{timetrap => 30_000},
+        begin
+            DB = ?FUNCTION_NAME,
+            {ok, _} = emqx_ds_new_streams:start_link(DB),
+            %% Create subscriptions:
+            {ok, Ref1} = emqx_ds_new_streams:watch(DB, [<<"foo">>, <<"1">>]),
+            {ok, Ref2} = emqx_ds_new_streams:watch(DB, [<<"foo">>, '+']),
+            {ok, Ref3} = emqx_ds_new_streams:watch(DB, [<<"foo">>, '#']),
+            {ok, Ref4} = emqx_ds_new_streams:watch(DB, [<<>>]),
```

Comment:
The output of `emqx_topic:words(<<>>)` is `['']` rather than `[<<>>]`, so using the former would be closer to real use cases.